increase prioritization fee (#50)
solana-foundation/solana-improvement-documents: proposals/0050-increase-prioritization-fee.md

---
simd: 0050
title: Increase Prioritization Fee
authors:
  - Tao Zhu
category: Standard
type: Core
status: Draft
created: 2023-05-22
feature: #31453
---

## Summary

Cost of prioritizing should be a meaningful portion of transaction base fee.

## Motivation

Prioritization fee is intended to be used sensitively during congestion,
provides advanced users a tool to access particular resources when contented
with a meaningful fee, while other users may opt to delay access to avoid

> paying extra fee. When used properly, it helps to reduce network congestion.

Reviewer: made an update, highlights "while other users may opt to delay access to avoid paying extra fee" via economic incentives. When account (or block) is contented, with local fee market, some users might decide isn't worth high prio fee to send their txs at that time, therefore helps to reduce congestion from sources.

> paying extra fee. When used properly, it helps to reduce network congestion.

Reviewer 2: why wouldn't they just not pay the fee, and send an unprioritized transaction?

> paying extra fee. When used properly, it helps to reduce network congestion.

Reviewer: Unprioritized transactions most likely will not have chance to be "tried" by leader, so not being able to add more contention to resources. quic would help on network spamming.

> paying extra fee. When used properly, it helps to reduce network congestion.

Reviewer 2: > Unprioritized transactions most likely will not have chance to be "tried" by leader

That really depends on how contested the accounts are. I guess the only reason they'd not submit is if they saw the **minimum** fee was > 0 for the account(s) or block.

> paying extra fee. When used properly, it helps to reduce network congestion.

Reviewer: > I guess the only reason they'd not submit is if they saw the minimum fee was > 0 for the account(s) or block.

yea, or similar scenario: https://github.com/solana-foundation/solana-improvement-documents/pull/50#issuecomment-1589485621

One fictional trading setup wrt priority fee could be: 
-  have one separate process keep pulling min prio fee for block and relevant account(s) from RPC endpoint, build local historical "fee market"s. 
- develop algos to utilize local historical data to predict prio fee for outgoing transactions. Perhaps something like "for Account `x`, with urgency of `y`, under market condition `z`, what should prioritization fee be?" 
- the trading process uses predicted prio fee for outbound transactions.

This requires the integrity of local fee market, which is part of the aiming of this proposal.

significantly more work on rolling out the change.

Current approach avoid changing API, nor introducing an additional terminology.

## Detailed Design

- No change to compute_budget instruction APIs;
- No change to how prioritization fee is calculated;
- No change to how banking_stage prioritization works;
- Add feature gated function to round user specified `compute_unit_price` down
  to its nearest 1_000 microlamports.
- The rounded `compute_unit_price` will be used by leader in prioritizing, and
  used by bank to calculate prioritization fee.

### Implementation

PoC https://github.com/solana-labs/solana/pull/31469

## Impact

- To vote transactions, no impact;
- To transactions don't set `compute_unit_price` (which is ~32% of non-vote
  transactions) , no impact;
- To transactions set `compute_unit_price`, user needs to reevaluate strategy
  of when to use priority, and how much. Specifically:
  - user might want to set `compute_unit_price` only when block or account
    contention increase;

    User can do that by either check RPC endpoint `getRecentPrioritizationFees`
    for minimal fee to land to block or write lock specific accounts when
    constructing transaction; 

>     User can also continuously pull RPC endpoint to built up prioritization

Reviewer: poll!

>     fee historical stats locally, then generate adquent prioritization fee

Reviewer 2: ```suggestion
    fee historical stats locally, then generate adequate prioritization fee
```

    algorithmically when constructing transactions.
  - when setting `compute_unit_price`, it is advised to set in increment of
    1_000 microlamports (eg 0.001 lamport);
  - When setting `compute_unit_price`, it is important to also evaluate
    value for `compute_unit_limit` to avoid paying too high prioritization fee;
    especially for ~50% non-vote transactions currently not setting
    `compute_unit_limit` at all.

### Examples

- 31.27% non-vote transactions in mainnet-beta set `compute_unit_price` when
  both block and account CUs are below 75% of limit (eg., not congested).

>   They are

Reviewer 2: weird line break

>   currently paying insignificant amount of prioritization fee; With this

Reviewer 2: ```suggestion
  currently paying an insignificant prioritization fee; With this
```

>   proposal, the payers of these transaction would have to consider only

Reviewer 2: ```suggestion
  proposal, the payers of these transactions would have to consider only
```

  paying prioritization fee when truly needed.

> - about 6% non-vote transactions in mainnet-beta that set `compute_unit_price`
>   but either not setting `compute_unit_limit` or setting it _above_ Max 1.4M.
>   with this proposal, payers of these transactions would have to consider
>   setting accurate/reasonable `compute_unit_limit`.

Reviewer 2: ```suggestion
- about 6% of non-vote transactions in mainnet-beta that set `compute_unit_price`
  are either not setting `compute_unit_limit` or setting it _above_ Max 1.4M.
  With this proposal, payers of these transactions would have to consider
  setting an accurate/reasonable `compute_unit_limit`.
```
